File: Editor/src/main/java/com/rspsi/studio/ui/DeclarativeToolUiRenderer.java
package com.rspsi.studio.ui;

import com.rspsi.editor.plugin.ui.EditorUiNode;
import com.rspsi.studio.theme.StudioPalette;
import com.rspsi.studio.theme.StudioWidgets;
import imgui.ImGui;
import imgui.flag.ImGuiSliderFlags;
import imgui.type.ImInt;

import java.util.List;

/**
 * Native Studio renderer for the frontend-neutral extension UI tree.
 *
 * <p>This class is intentionally one-way: extensions describe semantic widgets,
 * and Studio owns their actual ImGui projection, styling, IDs, focus behavior,
 * and future accessibility/scaling policy.</p>
 */
public final class DeclarativeToolUiRenderer {

    public void render(EditorUiNode node) {
        if (node != null) render(node, "root");
    }

    private void render(EditorUiNode node, String path) {
        switch (node) {
            case EditorUiNode.Text text -> {
                String value = text.value().get();
                if (text.muted()) ImGui.textDisabled(value == null ? "" : value);
                else ImGui.textWrapped(value == null ? "" : value);
            }
            case EditorUiNode.Button button -> {
                if (ImGui.button(button.label() + "##ext-" + path)) {
                    button.action().run();
                }
            }
            case EditorUiNode.Toggle toggle -> {
                boolean current = toggle.value().getAsBoolean();
                ImGui.alignTextToFramePadding();
                ImGui.text(toggle.label());
                ImGui.sameLine();
                boolean updated = StudioWidgets.toggleSwitch(
                        "ext-" + path, current, current ? "Enabled" : "Disabled");
                if (updated != current) toggle.onChange().accept(updated);
            }
            case EditorUiNode.IntSlider slider -> {
                ImGui.alignTextToFramePadding();
                ImGui.text(slider.label());
                ImGui.sameLine();
                int[] value = {slider.value().getAsInt()};
                ImGui.setNextItemWidth(Math.max(120.0f, ImGui.getContentRegionAvailX()));
                if (ImGui.sliderInt("##ext-" + path, value, slider.minimum(), slider.maximum())) {
                    slider.onChange().accept(value[0]);
                }
            }
            case EditorUiNode.DecimalSlider slider -> {
                ImGui.alignTextToFramePadding();
                ImGui.text(slider.label());
                ImGui.sameLine();
                float[] value = {(float) slider.value().getAsDouble()};
                ImGui.setNextItemWidth(Math.max(120.0f, ImGui.getContentRegionAvailX()));
                if (ImGui.sliderFloat(
                        "##ext-" + path,
                        value,
                        (float) slider.minimum(),
                        (float) slider.maximum(),
                        "%.2f",
                        ImGuiSliderFlags.None)) {
                    slider.onChange().accept(value[0]);
                }
            }
            case EditorUiNode.Select select -> {
                ImGui.alignTextToFramePadding();
                ImGui.text(select.label());
                ImGui.sameLine();
                List<String> options = select.options();
                String current = select.value().get();
                int currentIndex = Math.max(0, options.indexOf(current));
                ImInt selected = new ImInt(currentIndex);
                ImGui.setNextItemWidth(Math.max(120.0f, ImGui.getContentRegionAvailX()));
                if (ImGui.combo("##ext-" + path, selected, options.toArray(String[]::new))) {
                    select.onChange().accept(options.get(selected.get()));
                }
            }
            case EditorUiNode.Section section -> {
                ImGui.textColored(StudioPalette.ACCENT, section.title());
                ImGui.separator();
                renderChildren(section.children(), path + "-section");
            }
            case EditorUiNode.Row row -> renderRow(row.children(), path + "-row");
            case EditorUiNode.Column column -> renderChildren(column.children(), path + "-column");
            case EditorUiNode.Separator ignored -> ImGui.separator();
        }
    }

    private void renderChildren(List<EditorUiNode> children, String path) {
        for (int index = 0; index < children.size(); index++) {
            render(children.get(index), path + "-" + index);
        }
    }

    private void renderRow(List<EditorUiNode> children, String path) {
        for (int index = 0; index < children.size(); index++) {
            if (index > 0) ImGui.sameLine();
            render(children.get(index), path + "-" + index);
        }
    }
}

File: Editor/src/main/kotlin/com/rspsi/studio/ui/DeclarativeToolUiRenderer.kt
package com.rspsi.studio.ui

import com.rspsi.editor.plugin.ui.EditorUiNode
import com.rspsi.studio.theme.StudioPalette
import com.rspsi.studio.theme.StudioWidgets
import imgui.ImGui
import imgui.flag.ImGuiSliderFlags
import imgui.type.ImInt

/**
 * Native Studio renderer for the frontend-neutral extension UI tree.
 *
 * Extensions describe semantic widgets; Studio owns their ImGui projection,
 * styling, IDs, focus behavior, and future accessibility/scaling policy.
 */
class DeclarativeToolUiRenderer {
    fun render(node: EditorUiNode?) {
        if (node != null) {
            render(node, "root")
        }
    }

    private fun render(node: EditorUiNode, path: String) {
        when (node) {
            is EditorUiNode.Text -> {
                val value = node.value().get().orEmpty()
                if (node.muted()) {
                    ImGui.textDisabled(value)
                } else {
                    ImGui.textWrapped(value)
                }
            }

            is EditorUiNode.Button -> {
                if (ImGui.button("${node.label()}##ext-$path")) {
                    node.action().run()
                }
            }

            is EditorUiNode.Toggle -> {
                val current = node.value().asBoolean
                ImGui.alignTextToFramePadding()
                ImGui.text(node.label())
                ImGui.sameLine()

                val updated =
                    StudioWidgets.toggleSwitch(
                        "ext-$path",
                        current,
                        if (current) "Enabled" else "Disabled",
                    )
                if (updated != current) {
                    node.onChange().accept(updated)
                }
            }

            is EditorUiNode.IntSlider -> {
                ImGui.alignTextToFramePadding()
                ImGui.text(node.label())
                ImGui.sameLine()

                val value = intArrayOf(node.value().asInt)
                ImGui.setNextItemWidth(kotlin.math.max(MIN_CONTROL_WIDTH, ImGui.getContentRegionAvailX()))
                if (ImGui.sliderInt("##ext-$path", value, node.minimum(), node.maximum())) {
                    node.onChange().accept(value[0])
                }
            }

            is EditorUiNode.DecimalSlider -> {
                ImGui.alignTextToFramePadding()
                ImGui.text(node.label())
                ImGui.sameLine()

                val value = floatArrayOf(node.value().asDouble.toFloat())
                ImGui.setNextItemWidth(kotlin.math.max(MIN_CONTROL_WIDTH, ImGui.getContentRegionAvailX()))
                if (
                    ImGui.sliderFloat(
                        "##ext-$path",
                        value,
                        node.minimum().toFloat(),
                        node.maximum().toFloat(),
                        "%.2f",
                        ImGuiSliderFlags.None,
                    )
                ) {
                    node.onChange().accept(value[0].toDouble())
                }
            }

            is EditorUiNode.Select -> {
                ImGui.alignTextToFramePadding()
                ImGui.text(node.label())
                ImGui.sameLine()

                val options = node.options()
                val selected =
                    ImInt(
                        options
                            .indexOf(node.value().get())
                            .coerceAtLeast(0),
                    )
                ImGui.setNextItemWidth(kotlin.math.max(MIN_CONTROL_WIDTH, ImGui.getContentRegionAvailX()))
                if (ImGui.combo("##ext-$path", selected, options.toTypedArray())) {
                    node.onChange().accept(options[selected.get()])
                }
            }

            is EditorUiNode.Section -> {
                ImGui.textColored(StudioPalette.ACCENT, node.title())
                ImGui.separator()
                renderChildren(node.children(), "$path-section")
            }

            is EditorUiNode.Row -> renderRow(node.children(), "$path-row")
            is EditorUiNode.Column -> renderChildren(node.children(), "$path-column")
            is EditorUiNode.Separator -> ImGui.separator()
        }
    }

    private fun renderChildren(
        children: List<EditorUiNode>,
        path: String,
    ) {
        children.forEachIndexed { index, child ->
            render(child, "$path-$index")
        }
    }

    private fun renderRow(
        children: List<EditorUiNode>,
        path: String,
    ) {
        children.forEachIndexed { index, child ->
            if (index > 0) {
                ImGui.sameLine()
            }
            render(child, "$path-$index")
        }
    }

    companion object {
        private const val MIN_CONTROL_WIDTH = 120.0f
    }
}
